src/modules/roc_audio/profiling_reader.cpp

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_audio/profiling_reader.h"
+#include "roc_core/log.h"
+#include "roc_core/panic.h"
+
+namespace roc {
+namespace audio {
+
+ProfilingReader::ProfilingReader(IReader& reader,
+                                 core::IAllocator& allocator,
+                                 packet::channel_mask_t channels,
+                                 size_t sample_rate,
+                                 struct ProfilerConfig profiler_config)
+    : profiler_(allocator, channels, sample_rate, profiler_config)
+    , reader_(reader) {
+}
+
+bool ProfilingReader::read(Frame& frame) {
+    const core::nanoseconds_t elapsed = read_(frame);
+
+    bool ret = reader_.read(frame);
+    if (!ret) {
+        profiler_.add_frame(frame.size(), elapsed);
+    }
+    return ret;
+}
+
+core::nanoseconds_t ProfilingReader::read_(Frame& frame) {
+    const core::nanoseconds_t start = core::timestamp();
+
+    reader_.read(frame);
+
+    return core::timestamp() - start;
+}
```

Comment:
This code does something fundamentally wrong :)

What we do here:

* call reader_.read() once, and ignore its return value
* call reader_.read() once more (?) and store its return value
* if the second call failed (?), call add_frame()
* return stored return value from the second call...

What we should do instead:

1. call reader_.read() and store its return value
2. if the return value was true, i.e. if reader did the actual work, call add_frame()
3. return stored return value

The idea of (2) is to skip add_frame() call if read() did not perform any computations, so that its early and likely very fast return wouldn't affect our statistics. The idea of (3) is to forward return value from nested reader to the caller, so that the presence of the profiler in the pipeline would be transparent.

Sorry for not being clear first time.

What is really bad here is that we don't have pipeline integration tests that catch such a serious bug as calling read() twice instead of once(). That's because they don't enable profiler.

Those tests are located at `src/tests/roc_pipeline`. Could you please enabled profiling in them? (and ensure that they will catch this bug BTW).